### PR TITLE
Feature/integrate chronicle and opa tp

### PR DESCRIPTION
**Pull request by ryanbtp**
feat: Integrate chronicle and opa-tp

* Settings family reader
* From chain policy loading via settings  - policy details are not directly communicated
  with TP
* Ensure settings addresses are sent with tx
* CLI options - default behaviour for inmem is embedded, secure for production build
* API tests use simulated on-chain policies and settings
* Various log level adjustments
* Compose coordination - settings must run before chronicle tp + api

### CODE REVIEW

File: crates/chronicle-domain-test/src/.test.rs.pending-snap [hunk 1]
```diff
@@ -0,0 +1,546 @@
...ntext\n              }\n              instantActivity(id: {externalId: \"testactivity\"}) {\n                context\n              }\n              wasAssociatedWith(\n                responsible: { externalId: \"testagent\" }\n                activity: { externalId: \"testactivity\" }\n                role: MANUFACTURER\n              ) {\n                context\n              }\n            }\n            \"#)).await"},"snapshot":"{\n  \"data\": null,\n  \"errors\": [\n    {\n      \"message\": \"Api shut down before reply\",\n      \"locations\": [\n        {\n          \"line\": 3,\n          \"column\": 15\n        }\n      ],\n      \"path\": [\n        \"defineContractorAgent\"\n      ]\n    }\n  ]\n}"},"old":{"module_name":"chronicle_example__test","metadata":{},"snapshot":"{\n  \"data\": {\n    \"defineContractorAgent\": {\n      \"context\": \"chronicle:agent:testagent\"\n    },\n    \"defineItemManufacturedActivity\": {\n      \"context\": \"chronicle:activity:testactivity\"\n    },\n    \"instantActivity\": {\n      \"context\": \"chronicle:activity:testactivity\"\n    },\n    \"wasAssociatedWith\": {\n      \"context\": \"chronicle:agent:testagent\"\n    }\n  }\n}"}}
```
Comment (mtbc) [inline crates/chronicle-domain-test/src/.test.rs.pending-snap:1]: we can remove this file

File: crates/chronicle/src/bootstrap/mod.rs [hunk 2]
```diff
@@ -144,6 +106,7 @@ impl DatabaseConnector<(), StoreError> for RemoteDatabaseConnector {
     }
 }
 
+#[instrument(skip(db_uri))] //Do not log db_uri, as it can contain passwords
```
Comment (mtbc) [inline crates/chronicle/src/bootstrap/mod.rs:109]: Good catch.

File: docker/chronicle.yaml [hunk 3]
```diff
@@ -105,10 +107,12 @@ services:
       - PGPASSWORD=chronicle
     image: &chronicleImage chronicle-amd64:${ISOLATION_ID:-local}
     entrypoint: /bin/bash
+    links:
+      - "opa-init:opa-init"
     command:
       - -exc
       - |
-        sleep 10
+        sleep 120
```
Comment (mtbc) [inline docker/chronicle.yaml:115]: Maybe needn't be this long once we get things working?

Comment (ryanbtp) [reply docker/chronicle.yaml:120]: There's hopefully a better way to coordinate - it's quite fiddly but yes..

Comment (ryanbtp) [reply docker/chronicle.yaml:120]: We are using stop-on-container-exit to return test results, but we have an ephemeral init process, lots of fun

File: Makefile [hunk 4]
```diff
@@ -74,6 +74,8 @@ test-prep-$(ISOLATION_ID): $(HOST_ARCHITECTURE)-ensure-context
 		--builder ctx-$(ISOLATION_ID)-$(HOST_ARCHITECTURE) \
 		--platform linux/$(HOST_ARCHITECTURE) \
 		--target test \
+		--build-arg DOCKER_GID=$(DOCKER_GID) \
```
Comment (mtbc) [inline Makefile:77]: There's some rebase issue here, shouldn't need these.

File: docker/chronicle.yaml (hunk 3 above)
Comment (mtbc) [reply docker/chronicle.yaml:120]: May be worth a followup ticket.

File: Makefile [hunk 5]
```diff
@@ -123,6 +123,8 @@ $(1)-$(2)-build: $(2)-ensure-context  policies/bundle.tar.gz
 		--builder ctx-$(ISOLATION_ID)-$(2) \
 		--platform linux/$(2) \
 		--target $(1) \
+		--build-arg DOCKER_GID=$(DOCKER_GID) \
```
Comment (mtbc) [inline Makefile:126]: Shouldn't need these.

File: docs/opa.md [hunk 6]
```diff
@@ -312,3 +312,33 @@ arguments:
 ```bash
 opactl get-policy -i my_policy -o /path/to/output
 ```
+
+## Configuring chronicle to use OPA
```
Comment (mtbc) [inline docs/opa.md:316]:
```suggestion
## Configuring Chronicle to use OPA
```

File: docker/chronicle.yaml [hunk 7]
```diff
@@ -160,3 +173,29 @@ services:
...
+        /usr/local/bin/opactl generate --output ./root.pem
+        /usr/local/bin/opactl \
+          --sawtooth-address  tcp://validator:4004 \
+            bootstrap --root-key ./root.pem
+        /usr/local/bin/opactl \
+          --sawtooth-address tcp://validator:4004 \
+          set-policy \
+            --id allow_transactions -p /policies/bundle.tar.gz \
+            --root-key ./root.pem
+        while true; do sleep 30; done;
```
Comment (scealiontach) [inline docker/chronicle.yaml:199]: change this to a `tail -f /dev/null`

File: docker/chronicle.yaml [hunk 8]
```diff
@@ -64,21 +66,26 @@ services:
...
     image: &chronicleTpImage chronicle-tp-amd64:${ISOLATION_ID:-local}
-    entrypoint: >
-      /bin/bash -xc "/usr/local/bin/chronicle_sawtooth_tp
-      -C tcp://validator:4004
-      --console-logging pretty"
+    entrypoint: /bin/bash
+    command:
+      - -exc
+      - |
+        sleep 40
```
Comment (scealiontach) [inline docker/chronicle.yaml:76]: This one shouldn't really be necessary - if the validator isn't up the usual TP approach is to keep looping and retrying until it comes up

File: docker/chronicle.yaml [hunk 9]
```diff
@@ -99,16 +106,18 @@ services:
...
       - PGPASSWORD=chronicle
     image: &chronicleImage chronicle-amd64:${ISOLATION_ID:-local}
     entrypoint: /bin/bash
+    links:
+      - "opa-init:opa-init"
     command:
       - -exc
       - |
-        sleep 10
+        sleep 60
```
Comment (scealiontach) [inline docker/chronicle.yaml:120]: this one would be better as a `wait-for-it` and then *maybe* an extra sleep.  I thought we had already put that in somewhere, no?

File: docker/chronicle.yaml [hunk 10]
```diff
@@ -160,3 +173,29 @@ services:
...
+    entrypoint: /bin/bash
+    environment:
+      - RUST_BACKTRACE=full
+      - RUST_LOG=info,cranelift_codegen=off,wasmtime_cranelift=off,regalloc2::ion=off,cranelift_wasm=off
+    volumes:
+      - ${PWD}/policies:/policies
+    command:
+      - -exc
+      - |
+        sleep 20
```
Comment (scealiontach) [inline docker/chronicle.yaml:188]: this should really be `wait-for-it`  and then *maybe* a sleep.

**Review by suchapalaver — APPROVED** [review]

File: docker/chronicle.yaml (hunk 9 above)
Comment (ryanbtp) [reply docker/chronicle.yaml]: We don't bake wait for it into the chronicle image, would need a derived container or test dependencies in released artefacts

**Comment by mtbc** [comment]
This seems to be working just fine with `make run` and my time-recording example domain.

File: crates/api/src/lib.rs [hunk 11]
```diff
@@ -1548,8 +1558,62 @@ mod test {
...
+        .write_to_vec(&mut buf)
+        .unwrap();
 
-        EmbeddedChronicleTp::new()
+        let setting_entrypoint = (
+            chronicle_protocol::settings::sawtooth_settings_address("chronicle.opa.entrypoint"),
+            buf,
+        );
+
+        let d = env!("CARGO_MANIFEST_DIR").to_owned() + "/../../policies/bundle.tar.gz";
```
Comment (mtbc) [inline crates/api/src/lib.rs:1595]: Curious about the tradeoff here with using rust-embed, anything to know or worry about?

File: crates/chronicle-domain-test/src/test.rs [hunk 12]
```diff
@@ -127,7 +131,63 @@ mod test {
         let keystore = DirectoryStoredKeys::new(keystore_path).unwrap();
         keystore.generate_chronicle().unwrap();
 
-        let tp = EmbeddedChronicleTp::new();
+        let mut buf = vec![];
+        Setting {
+            entries: vec![Setting_Entry {
+                key: "chronicle.opa.policy_name".to_string(),
```
Comment (mtbc) [inline crates/chronicle-domain-test/src/test.rs:137]: Where we're reusing a lot of these strings I wonder if it's worth defining them as constants somewhere (perhaps in some settings enum).

File: crates/chronicle/src/bootstrap/mod.rs [hunk 13]
```diff
@@ -264,6 +232,60 @@ fn construct_db_uri(matches: &ArgMatches) -> String {
...
+        let opa = opa_executor_from_embedded_policy(default_policy_name, entrypoint).await?;
+        tracing::warn!(
+            "Chronicle operating in an insecure mode with an embedded default OPA policy"
+        );
+        Ok(ConfiguredOpa::Embedded(opa))
+    } else {
+        let (opa, settings) =
+            self::opa::opa_executor_from_sawtooth_settings(&sawtooth_address(config, options)?)
+                .await?;
+        tracing::info!(use_on_chain_opa= ?settings, "Chronicle operating in secure mode with on chain OPA policy");
```
Comment (mtbc) [inline crates/chronicle/src/bootstrap/mod.rs:284]:
```suggestion
        tracing::info!(use_on_chain_opa= ?settings, "Chronicle operating in secure mode with on-chain OPA policy");
```

File: crates/sawtooth-tp/src/opa.rs [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    ) -> Result<ExecutorContext, ApplyError> {
+        let policy_name_settings_entry =
+            ctx.get_state_entry(&sawtooth_settings_address("chronicle.opa.policy_name"))?;
+
+        let entrypoint_settings_entry =
+            ctx.get_state_entry(&sawtooth_settings_address("chronicle.opa.entrypoint"))?;
+
+        match (policy_name_settings_entry, entrypoint_settings_entry) {
+            (None, None) => {
+                warn!("Insecure operating mode - no on chain policy name or entrypoint settings found");
```
Comment (mtbc) [inline crates/sawtooth-tp/src/opa.rs:46]:
```suggestion
                warn!("Insecure operating mode - no on-chain policy name or entrypoint settings found");
```

File: crates/sawtooth-tp/src/opa.rs [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+                Ok(self.embedded.clone())
+            }
+            (Some(policy_name_settings_entry), Some(entrypoint_settings_entry)) => {
+                info!("Chronicle operating in secure mode - on chain policy name and entrypoint settings found");
```
Comment (mtbc) [inline crates/sawtooth-tp/src/opa.rs:50]:
```suggestion
                info!("Chronicle operating in secure mode - on-chain policy name and entrypoint settings found");
```

File: crates/sawtooth-tp/src/opa.rs [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+                let policy_name_settings_entry: Setting =
+                    Message::parse_from_bytes(&policy_name_settings_entry).map_err(|_e| {
```
Comment (mtbc) [inline crates/sawtooth-tp/src/opa.rs:52]: With these various internal errors, ought we helpfully log the entry and (where there is one) the underlying error?

File: docs/opa.md [hunk 17]
```diff
@@ -336,3 +336,33 @@ arguments:
...
+
+### Load a policy bundle using opactl
+
+The policy id here must match the rego.
+
+```bash
+opactl set-policy -i allow_transactions -p /path/to/bundle.tar.gz
+```
+
+### Configure sawtooth with settings that match the policy
```
Comment (mtbc) [inline docs/opa.md:353]:
```suggestion
### Configure Sawtooth with settings that match the policy
```

**Comment by mtbc** [comment]
Latest batch of questions are probably easily answered or resolved, no showstoppers from my side.

File: crates/api/src/lib.rs (hunk 11 above)
Comment (ryanbtp) [reply crates/api/src/lib.rs:1595]: It's test only, and I was potentially considering scenarios where we use
different policies to the ones we embed to simulate production behaviour where
you can use non default policies

Comment (mtbc) [reply crates/api/src/lib.rs:1595]: That sounds to be a useful developer tip!

File: crates/chronicle/src/bootstrap/mod.rs (hunk 13 above)
Comment (mtbc) [reply crates/chronicle/src/bootstrap/mod.rs:284]: This is still not fixed but can approve without if expedient.

**Review by mtbc — APPROVED** [review]
No outstanding release blockers.

**Pull request merged**